### PR TITLE
[Pipeline Middleware] fix data race in Pipeline Scheduler for DAG

### PR DESCRIPTION
**Pull request by Wesley-Jzy**
Fix the data race introduced by adding DAG Topo for scheduler.

### CODE REVIEW

File: tests/test_pipeline/test_middleware_1f1b.py [hunk 1]
```diff
@@ -42,24 +41,44 @@ def run_master(args):
...
+    data_kwargs = [
+        data_gen('MLP'),
+        data_gen('DAG_MLP'),
+    ]
+    model = [
+        MLP(dim, stage_num * 3),
+        DAG_MLP(dim, stage_num * 3),
+    ]
+
+    i = 1
```
Comment (YuliangLiu0306) [inline tests/test_pipeline/test_middleware_1f1b.py:66]: I think using parameterize decorator in colossalai.testing is a more elegant implementation.
For example:
@parameterize('model_cls', [MLP, DAG_MLP])
def test(model_cls):

File: colossalai/pipeline/rpc/_pipeline_base.py [hunk 2]
```diff
@@ -360,11 +363,15 @@ def subscribe_producer(self, microbatch_id: int, forward_only: bool):
...
+                
+        return work_item_from_producer
+    
+    # TODO(jiangziyue) Profile the side effect of the lock for lifecycle protection and consider a better one.
+    def subscribe_producer(self, microbatch_id: int, forward_only: bool):
+        key = UniqueKey(microbatch_id, Phase.FORWARD)
         with self.work_list_condition_lock:
-            key = UniqueKey(microbatch_id, Phase.FORWARD)
             if key not in self.work_list:
+                work_item_from_producer = self._subscribe_producer(microbatch_id, forward_only)
```
Comment (YuliangLiu0306) [inline colossalai/pipeline/rpc/_pipeline_base.py:374]: For the maintainability, a more detail comment is expected here to illustrate why we put these actions under the lock context.

File: colossalai/pipeline/rpc/_pipeline_base.py [hunk 3]
```diff
@@ -188,23 +184,37 @@ def sync_global_worker_rrefs(self, pp_rank_to_worker_rref: Dict[int, PyRRef]) ->
 
     def get_output_by_key(self, key: UniqueKey, recv_rank=None) -> Any:
         with self.output_list_condition_lock:
+            #print(f'rank {self.pp_rank} waiting for {key} to recv rank {recv_rank}')
```
Comment (YuliangLiu0306) [inline colossalai/pipeline/rpc/_pipeline_base.py:187]: Remove these commented lines.

File: colossalai/pipeline/rpc/_pipeline_base.py (hunk 2 above)
Comment (Wesley-Jzy) [reply colossalai/pipeline/rpc/_pipeline_base.py:374]: Done.

**Review by YuliangLiu0306 — APPROVED** [review]

**Comment by Wesley-Jzy** [comment]
image

**Comment by Wesley-Jzy** [comment]
image

**Pull request merged**